python/taichi/lang/ops.py

```diff
@@ -55,6 +56,27 @@ def wrapped(a, b):
     return wrapped
 
 
+inplace_binary_ops = []
+
+
+def inplace_binary(foo):
```

Comment:
I think this is very tricky and needs very careful thinking.

Here, the inplace-ness comes from the atomics operations, **not** from `element_wise_inplace_binary()`. Assuming my previous comment has been resolved:

1. If we still use `element_wise_inplace_binary` here, then we are basically doing `a[i] = ti.atomic_op(a[i], b); return a`. So this is wrong, because the atomic operation result on `a` are lost.
2. Therefore, we need to use `element_wise_binary` --> `ret = ti.Matrix(a.n, a.m); ret[i] = ti.atomic_op(a[i], b); return ret`. Then we are returning a new variable `ret`, while `a` preserves the atomic operations.

Unfortunately, I think atomic operations are very special, in that it is an operation **with side effect**. What this means is that, if you combine `element_wise_inplace_binary` with an atomic op, then this is completely broken, due to the reason in 1. (It's not broken now because `element_wise_inplace_binary` is not really inplace..)

So, my conclusion & suggestion is to add a special `element_wise_atomic_binary` to `ti.Matrix`. And use that in this decorator. Note that the decorator's name should also be change to reflect the fact that it is only meant for decorating atomic operations.

---

On an unrelated topics, can we stop using `foo` and give it a more meaningful name, e.g. `op`? I know this is not your fault, since it is a legacy problem, but we can at least stop widening this problem...

Reply:
tl;dr, maybe you want the name nit `atomic_binary`?

Reply:
Good point. If we don't need that at all, maybe we can use an even more generic name, e.g. `broadcast_binary` or something? Do we still need `element_wise_atomic_binary`? Seems like `element_wise_binary` is enough?

Also, change `foo` to a better name?

Reply:
IMO `op` may make people think of `'Add'` in `augassign` instead of a function, wab `op_func`? Anyway, ff2 refactor these name nits iapr on your own, I'll keep `foo` here as-is to prevent conflict with #1051 and #1046 (in fact, there's already some conflicts...)

Reply:
I see. Then what about the `atomic_binary` stuff?

(One thing i should point out... Acronyms like BTW and FYI are very common and widely adopted in the English community, whereas wab and iapr don't really much make sense and actually hinder the communication. I found this https://www.smart-words.org/abbreviations/text.html, in case you are interested.)

Reply:
This table looks cool to me! But IMO it could be great if we could have an acronym table just for us as GitHub user :)
1. iapr, in another PR
2. itpr, in this PR
3. ff2, free free to
4. mt, merge this
5. pvc, prevent conflict
6.LGTMig, LGTM in general (except for a few nits)
Sadly, they're not widely adopted in generic English...at all.

Reply:
> broadcast_binary

How about `writeback_binary`? We need to distinguish `writeback_binary` and `binary` since we want to enforce that:
```py
scalar += matrix  # error
scalar + matrix   # OK
```

Reply:
> Sadly, they're not widely adopted in generic English...at all.

Yeah. AFAICT, the only acronym in the coding world that I can think of is LGTM..

I pointed this out because, if we started to use these customized acronyms, it would soon become jargons that are only known by the Taichi devs. Then it will prevent new Taichi contributors from understanding our discussion... So let's stick with the standard acronyms.